Add option for static fallback of schema URI

danielkza opened the pull request.

danielkza committed:
Add option for static fallback of schema URI

ottomata commented:
Hey thanks!  If I understand correctly, the intention here is to be able to support a single manually provided schema URI for all messages?

I think this is fine!  BTW though, this connector as is is still pretty WIP, and we aren't yet using it at the Wikimedia Foundation.  I'm still hoping to sometime in the next 4-6 months, but we will see how things go.  When that happens, we'll likely move this repo over to the wikimedia/ github namespace, and some changes will be made then too.  E.g., perhaps is would be better to have a configurable SchemaLoader interface, with a couple of implementations here.  One that gets the schema from a URI in an event, another that provides the schema explicitly for all events, another that provides a static URI for all events (like your patch here does), etc.

If you're ok with a lot of this changing later, I think we can merge this in here now.

danielkza commented:
> Hey thanks! If I understand correctly, the intention here is to be able to support a single manually provided schema URI for all messages?

Yes! I have a bunch of topics in JSON format I want to support that do not have schemas now, and I would still like to process the existing data.

> I think this is fine! BTW though, this connector as is is still pretty WIP, and we aren't yet using it at the Wikimedia Foundation. ...

That's not a problem, I'm currently working on deploying it in Docker images, so if I hit a breaking change I can just hold back to an older commit. Also, unlike you guys I don't have plans to use JSONSchema in the long term; it's just an interim solution before eventually moving to a structured format for message production (like Avro or Protobuf).

> If you're ok with a lot of this changing later, I think we can merge this in here now.

I read the fine print, no worries :)

I just realized it might be helpful for (even myself) to be able to replace a placeholder for the topic name in the fallback (so I can use the same connector for multiple topics. I think I'll add that as well, so hold on a little bit before merging please.

ottomata commented:
Hm, does your data not have schemas at all, or you mean they just don't have schema URI pointers in the data?

If no schemas at all, you can just use Kafka Connect in schemaless mode and treat the JSON as a string.  Or, Ewen (from Confluent) has a Single Message Transform that will infer and set a Connect schema from each JSON message:  https://github.com/apache/kafka/blob/caca9101b87ca78368ece3e57c1719d4465ae16f/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchema.java

You can also find some discussion on how to use this in this thread: https://groups.google.com/forum/#!msg/confluent-platform/OGuOCkEKoeA/CH3X3E3ABwAJ

danielkza commented:
> Hm, does your data not have schemas at all, or you mean they just don't have schema URI pointers in the data?

There are no pointers, but the data in each topic is consistent. I created schemas for each of them externally and plan on adding them to the Kafka Connect images.

> If no schemas at all, you can just use Kafka Connect in schemaless mode and treat the JSON as a string. Or, Ewen (from Confluent) has a Single Message Transform that will infer and set a Connect schema from each JSON message: https://github.com/apache/kafka/blob/caca9101b87ca78368ece3e57c1719d4465ae16f/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchema.java

I looked into that, but even when inferring the schemas (which I actually did), it seems better to me to do it once and have it persisted somewhere, so that I can catch any unwanted incompatibilities and use it as a reference for future format changes.

danielkza committed:
Add Apache Commons Lang dependency

It was already pulled transitively by other dependencies, but add it
directly so we can start to use it in a proper way.

danielkza committed:
Allow topic name placeholder substitution in schema URI fallback

hashhar commented:
>perhaps is would be better to have a configurable SchemaLoader interface, with a couple of implementations here. One that gets the schema from a URI in an event, another that provides the schema explicitly for all events, another that provides a static URI for all events (like your patch here does), etc.

@ottomata Regarding this, do you have any plans at Wikimedia on working in this direction? If yes then I'd be happy to contribute some code that I wrote that does exactly what you say.

If there are no plans to continue with this converter then is it fine if I make my own converter public which is very similar to this one (had to reimplement it all from scratch since there is no license in this repo)?

ottomata commented:
Unfortunately, since Confluent changed their licenses to their non-free-open-source Confluent Community License, the Wikimedia Foundation can't make use of most of the good Connectors out there.  Our main use case was to use Kafka Connect HDFS for Hadoop and Hive integration, but now we aren't sure what to do.  We're still stuck on Camus, and are considering Gobblin.

We aren't likely to work on this in the future.

hashhar commented:
We have decided to stay on the 5.1 branch (it's Apache 2.0) too due to this reason while our legal team irons out all the details. It's sad but unfortunately a little necessary given the past behavior of cloud providers.

ottomata commented:
As for License for this repo, sorry!  This was just a WIP until we moved under wikimedia's github account.  I'll add a License and you can feel free to fork if you wish.
